### PR TITLE
Delegation

### PR DESCRIPTION
**Pull request by hansl**
Allow servers to handle delegation of identities in messages, through a certification process.

As part of this specification directly, a lot of additional cracks in the specification were filled, including how to properly resolve an identity from a message, creating addresses from public keys in a canonical way, and some other quality of life improvement to readability and typing.

### CODE REVIEW

**Comment by hansl** [comment]
@fmorency This is ready to review. Will update the description.

File: attributes/response/cddl/2_delegation.cddl [hunk 1]
```diff
@@ -0,0 +1 @@
+delegation@response-attribute-arg = delegation-certificate / [ + delegation-certificate ]
```
Comment (fmorency) [inline attributes/response/cddl/2_delegation.cddl:1]: I'll comment here only once, but we MIGHT want to put an upper limit on the certification chain.

I wonder what's the performance impact on the server of maxing a block size with a very long certificate chain.

That makes me think we should probably create a "performance" test suite with those edge cases.

**Review by fmorency — CHANGES_REQUESTED** [review]
First iteration! Thanks!

Some comments
- Remove useless wording. Keep the minimum
- Multiple short phrases are better than a single long one
- Flows are better explained using diagrams and text
- You replaced a bunch of `identity` with `address`, but a lot of `identity` are still present, e.g., in `7_kvstore_commands.cddl`. Please fix them all.
- Vocabulary needs to be consistent, i.e., ID, address, identity, ...

Some questions
- Can we revoke a delegation?
- Are we planning on being able to limit what kind of operation the key is able to perform? E.g., the key pair is able to sign/encrypt, but we want to limit delegation to signing only.
- What are the main use cases you want to address with this feature?
-

File: spec/protocol/delegation.adoc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+The MANY protocol relies on digital signatures of messages to identify users and servers.
+The resolution of the address is defined in the xref:envelope.adoc#_identity[Envelope – Identity] section.
+
+This specification adds and describe an additional attribute and resolution step; delegation.
+Delegation happens when an identity signs a delegation certificate to another identity, allowing the private key of the latter one to sign messages where the sender is the first one.
+
+== Overview
+
+Given the identities `Alice` and `Bob`, assuming `Alice` would like to allow `Bob` to impersonate them in a message.
+`Alice` would sign a delegate certificate to `Bob`, which `Bob` would include in their messages.
```
[the remark this answers is not in the document]
Comment (hansl) [reply spec/protocol/delegation.adoc:18]: Yeah there's no need to be specific.

File: spec/protocol/identity.adoc [hunk 3]
```diff
@@ -1,19 +1,28 @@
 = Identity
 
 Every actor, signer and networks have an identity (materialized in a ManyVerse Address), which is similar to an address in Ethereum.
+This identity is usually represented by a private and public key pair.
```
[the remark this answers is not in the document]
Comment (hansl) [reply spec/protocol/identity.adoc:4]: I've made the change, but I don't like restricting language in general. Maybe we won't use a private/public key pair in the future (with symmetric keys for example), or maybe identity won't be about key pair (a did URL for example).

File: attributes/network/10_delegation.adoc [hunk 4]
```diff
@@ -0,0 +1,33 @@
+= Delegation Attribute (#10)
```
[the remark this answers is not in the document]
Comment (hansl) [reply attributes/network/10_delegation.adoc:1]: Not sure what that means but if it means having the endpoints as second level headers, sure.

File: spec/protocol/delegation.adoc [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      . if the argument is a single certificate, perform step 2 and 3 once for it.
+
+2. for each certificate, validate (in order);
+   a. the signature using the same <<./envelope.adoc#_identity,Identity Resolution>> steps for a regular envelope,
+   b. the `to` (1) field is the same as the identity resolved thus far (either the last identity resolved as part of this step, or the identity resolved from the message envelope),
+   c. the `expiration` (2) field is still in the future (must also not be equal).
+   d. if the `final` (4) field is `true`, and there are more certificates to evaluate, the server MUST return an error.
+
+3. if all validation above pass, the latest identity is the `from` (0) field of the certificate.
+   If there are more certificate to resolve, go back to step 2 with the next certificate.
```
[the remark this answers is not in the document]
Comment (hansl) [reply spec/protocol/delegation.adoc:90]: I removed the last line since at step 2 I already say "for each certificate". I could provide pseudocode but I don't think a diagram would be easy to make in a way that makes it also useful.

File: spec/protocol/envelope.adoc [hunk 6]
```diff
@@ -0,0 +1,33 @@
+= MANY Envelope Specification
+:cddl: ./cddl/
+// Metadata
+:toc:
+:hide-uri-scheme:
+
+== Introduction
+
+The MANY protocol relies on digital signatures of messages to identify users and servers.
+Messages are tagged COSE Sign envelopes containing headers and a signature (and optionally counter signatures).
```
[the remark this answers is not in the document]
Comment (hansl) [reply spec/protocol/envelope.adoc:10]: Not yet. I'll remove that mention. We will though.

File: spec/protocol/identity.adoc [hunk 7]
```diff
@@ -1,19 +1,28 @@
...
+
+2. [[public_key_address]]*Public Key.*
+   A public key can be a program or a user that is the source of a message and can send messages.
+   It MUST sign all messages sent, and its ID is a hash of the public key that can validate the signature.
+
 3. *Subresource.*
-A Subresource address is an indexed resource inside a network that does not have its own internal public key but can be addressed.
-Messages to and from a subresource should use the same public key as the network they're part of.
+   A Subresource address is an indexed resource inside a network that does not have its own internal public key but can be addressed.
+   Messages to and from a subresource should use the same public key as the network they're part of.
```
[the remark this answers is not in the document]
Comment (hansl) [reply spec/protocol/identity.adoc:22]: We do when we evaluate the key itself. `matches()` is the method that does the check.

File: spec/protocol/attributes.adoc [hunk 8]
```diff
@@ -32,3 +42,19 @@ A CBOR tag on the payload value (not in the bytestring) must be added by those p
...
+
+== Unknown Attributes
+
+A server who receives a request with an attribute it does not recognize MUST automatically return an error with code `UnknownRequestAttribute` (-12) and a list of attributes that caused the error.
+No further action should be taken; an unknown attribute should have no side effect.
+The client is then welcome to resend the request without the faulty attribute.
+
+A client receiving a response with an attribute it does not recognize should ignore the attribute, and likely report a warning.
+In some cases the payload might change, so a client that sees unknown attributes in a response should try its best to interpret its payload.
+As an example, a CLI tool that sees `[1, ...]` attribute and doesn't support asynchronous responses should error out with a comment that the payload was empty (see xref:../../attributes/response/1_async.adoc[Response Attribute 1 - Async] for more details).
```
[the remark this answers is not in the document]
Comment (hansl) [reply spec/protocol/attributes.adoc:60]: Because you're in a situation where you sent a request, the request executed an action, you received a valid response, but there are attributes in the response you do not understand. An error might imply the user should re-execute the message. But the message was executed. You're having an error _after the fact_. The message would be like "Error: Success".

File: attributes/response/cddl/2_delegation.cddl (hunk 1 above)
Comment (hansl) [reply attributes/response/cddl/2_delegation.cddl:1]: Yeah I'll leave this to servers. There is an upper side limit to the amount of certificates (message size) but it does take time to process each of them so there's some DoS possible.

**Comment by hansl** [comment]
> Can we revoke a delegation?

No. This would require central infrastructure (PKI) that we are not planning to build.

> Are we planning on being able to limit what kind of operation the key is able to perform? E.g., the key pair is able to sign/encrypt, but we want to limit delegation to signing only.
What are the main use cases you want to address with this feature?

Not right now. We could extend the certificate for this kind of thing.

File: spec/protocol/attributes.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (fmorency) [reply spec/protocol/attributes.adoc:59]: Make sense. It wasn't clear in the spec; I would describe it like you just did.

File: spec/cddl/delegation.cddl [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+   ; An expiration timestamp. If the system time is past this timestamp,
+   ; the certificate is invalid and the server MUST return an error without
+   ; opening the envelope further.
+   2 => time,
+
+   ; A description or message about this certificate.
+   ? 3 => memo,
+
+   ; Whether or not this MUST be the final certificate (no sub-delegation
+   ; allowed).
```
[the remark this answers is not in the document]
Comment (fmorency) [reply spec/cddl/delegation.cddl:20]: I'm okay with it, as long as `2 => time` is mandatory.

File: spec/protocol/delegation.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (fmorency) [reply spec/protocol/delegation.adoc]: Pseudocode would be a good compromise.

File: spec/protocol/envelope.adoc [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+3. _This is where additional resolution steps should be inserted._
+   If the server does not support other resolution steps, it should pass to the next step.
+
+4. The resulting key is compared with the message's `from` field:
+   a. if the `from` address is anonymous and the resolved key is not, the server MUST return an error.
+   b. if the `from` address is a public key and it matches the resolved key, the address is used as the sender.
+   c. if the `from` address is a sub-resource identity and its public key matches the resolved key, use the subresource identity address as the sender.
+
+The final resolved address can be used to identify the sender where it makes sense (e.g. resolving ACLs).
```
[the remark this answers is not in the document]
Comment (fmorency) [reply spec/protocol/envelope.adoc]: Same compromise on pseudocode.

File: spec/protocol/identity.adoc [hunk 11]
```diff
@@ -68,16 +78,26 @@ Two characters give us 10 bits of entropy, which indicates that two typos on a t
...
+
+== COSE Key
+
+Resolving the address of a COSE Key is simple;
+
+1. Operations MUST NOT contain the `Sign`, `Encrypt`, `Wrap` or `Derive` operations.
+2. Any private key parts (e.g. the `D` point in elliptic curves) must be missing.
+3. Serialize the COSE Key in its CBOR binary format, using the Canonical CBOR encoding rules (see https://www.rfc-editor.org/rfc/rfc7049#section-3.9[Canonical CBOR]).
+4. Calculate the SHA3-224 hash of the resulting blob.
+5. Generate a link:#public_key_address[Public Key] Address with the `Key` part the result of the SHA3-224.
```
[the remark this answers is not in the document]
Comment (fmorency) [reply spec/protocol/identity.adoc:102]: Same compromise on pseudocode.

File: attributes/network/cddl/10_delegation.cddl [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+delegate.createCertificate@param = {}
+
+; Returns a delegation.
+delegate.createCertificate@return = {
+    0 => delegation-certificate,
+}
+; end::create-certificate[]
+
+; tag::whoami[]
+; Returns which identity the message to. Useful for testing delegations.
```
Comment (fmorency) [inline attributes/network/cddl/10_delegation.cddl:12]:
```suggestion
; Returns which address the message resolves to. Useful for testing delegations.
```

**Review by fmorency — CHANGES_REQUESTED** [review]
Lgtm. Small things to fix and I think we can merge afterward. Thanks!

- `3_kvstore.cddl` still refers to `non-anonymous-identity`
- More references to `identity` to fix in `adoc` files
- Add pseudocode
- Rework the `error after the fact` section

File: spec/protocol/attributes.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (hansl) [reply spec/protocol/attributes.adoc:59]: Added more text.

**Comment by hansl** [comment]
PTAL.

**Review by fmorency — APPROVED** [review]
Lgtm assuming comments are addressed.
